File: app/src/main/java/com/shrikantbadwaik/tmdb/domain/usecase/ApiUseCase02.java
package com.shrikantbadwaik.tmdb.domain.usecase;

import com.shrikantbadwaik.tmdb.data.remote.CallbackWrapper;

import retrofit2.Call;

public abstract class ApiUseCase02<IN, OUT> {

    public ApiUseCase02() {
    }

    public void execute(IN input, CallbackWrapper<OUT> callbackWrapper) {
        prepareObservableUseCase(input).enqueue(callbackWrapper);
    }

    private Call<OUT> prepareObservableUseCase(IN input) {
        return buildObservableUseCase(input);
    }

    protected abstract Call<OUT> buildObservableUseCase(IN input);
}

File: app/src/main/java/com/shrikantbadwaik/tmdb/domain/usecase/ApiUseCase02.kt
package com.shrikantbadwaik.tmdb.domain.usecase

import com.shrikantbadwaik.tmdb.data.remote.CallbackObserverWrapper
import com.shrikantbadwaik.tmdb.domain.helper.rx.SchedulerProvider
import io.reactivex.Observable
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.observers.DisposableObserver

abstract class ApiUseCase02<INPUT, RESULT>(
    private val schedulerProvider: SchedulerProvider,
    private val compositeDisposable: CompositeDisposable = CompositeDisposable()
) {
    fun execute(input: INPUT, callbackWrapper: CallbackObserverWrapper<RESULT>) {
        val disposable: DisposableObserver<RESULT> = prepareObservableUseCase(input).subscribeWith(callbackWrapper)
        compositeDisposable.add(disposable)
    }

    private fun prepareObservableUseCase(input: INPUT): Observable<RESULT> {
        return buildObservableUseCase(input).subscribeOn(schedulerProvider.io).observeOn(schedulerProvider.ui)
    }

    protected abstract fun buildObservableUseCase(input: INPUT): Observable<RESULT>
}